geozero/src/geometry_processor.rs

```diff
@@ -186,6 +246,16 @@ pub trait GeomProcessor {
     }
 
     /// Begin of `GeometryCollection` processing
+    ///
+    /// ## Parameters
+    ///
+    /// - `size`: the number of geometries in this GeometryCollection.
+    /// - `idx`: the positional index of this GeometryCollection. Unsure when this is non-zero?? Nested geometry collections?
```

Comment:
GeoZero has FeatureProcessing and GeometryProcessing apis and the interaction between the two can be ill defined at times.

For example, a format like geojson supports the _feature_ processing APIs, allowing you to enumerate each feature's geometry and properties. But geojson also supports the Geometry-only processing apis. So you can iterate over each feature, but only process the geometry - ignoring its properties.

In this case, the `n`th feature of a feature collection will have a geometry with index `n`.
See https://github.com/georust/geozero/blob/main/geozero/src/geojson/geojson_reader.rs#L140

A related issue is https://github.com/georust/geozero/pull/35, wherein processing feature collections as "geometry only" produces invalid output for some formats.

Reply:
I'm not sure if I fully understood that. Does that mean that the text I added for `multipolygon_begin` is wrong:

> - `idx`: the positional index of this MultiPolygon. This will be 0 except in the case of a GeometryCollection.

I.e. is that idx the row of the dataset?

Reply:
Maybe this example will clarify:

```
{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "geometry": {
        "type": "Point",
        "coordinates": [ -18, 44 ]
      },
      "properties": {
        "name": "Point A",
        "description": "This is the first point."
      }
    },
    {
      "type": "Feature",
      "geometry": {
        "type": "Point",
        "coordinates": [ 71, 58 ]
      },
      "properties": {
        "name": "Point B",
        "description": "This is the second point."
      }
    }
  ]
}
```

When you process this with a FeatureProcessor, you'll get something like:
```
begin_dataset
  begin_feature 0 
    begin_geometry 0
    end_geometry 0
  end_feature 0
  begin_feature 1
    begin_geometry 0
    end_geometry 0
  end_feature 1
end_dataset
```

Notice that in the above, the args to `begin_geometry` are always 0.

If you process this with GeometryProcessor, I think you'll get something different:
```
begin_dataset
  begin_geometry 0
  end_geometry 0
  begin_geometry 1
  end_geometry 1
end_dataset
```

Notice the args to `begin_geometry` are _not_ always 0 in this example.

I think both of these behaviors make sense, but it's a little tricky.

Reply:
I kinda interpret that as "begin_geometry passes the `idx` of the geometry". When passed via `FeatureProcessor`, each row is its own geometry and so the `idx` is always `0`. When passed via `GeometryProcessor`, all rows are considered as one `GeometryCollection`, and so each row has its own `idx`

Reply:
That interpretation makes sense to me. So in response to your comment in the code:

> Unsure when this is non-zero?? Nested geometry collections?

1. yes, nested geometry collections
2. but also when you're using GeometryProcessor to process a _Feature_ whose geometry is a GeometryCollection. As in:
```
{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "geometry": {
        "type": "GeometryCollection",
        "geometries": [ ... ]
      },
      "properties": {
        "name": "Feature A",
        "description": "This is the first feature."
      }
    },
    {
      "type": "Feature",
      "geometry": {
        "type": "GeometryCollection",
        "geometries": [ ... ]
      },
      "properties": {
        "name": "Feature B",
        "description": "This is the second feature."
      }
    }
  ]
}
```

begin_geometry will be called with a non-zero index when GeometryProcessing gets to "Feature B".

Reply:
Every time I come back to this PR I'm tired and my mind hurts trying to understand this 🙃 . How do you like the rewording I just added?

Reply:
Perfecto. Sorry about the brain pain.